### DISCUSSION TITLE
kube-bind does not support cluster-scoped resources - how to cope?

### DISCUSSION THREAD

**Discussion by MikeSpreitzer** [kubestellar/kubestellar#582] [General]
[kube-bind](https://github.com/kube-bind/kube-bind) does not support non-namespaced resources.  Today KubeStellar relies on the fact that kcp's APIExport/APIBinding supports cluster-scoped resources; following are the particular dependencies that involve non-namespaced resources.

1. The KubeStellar API (the edge.kcp.io API group).
2. The APIExports for SyncTarget and Location.
3. The APIExport for Workspace.
4. Not today, but in future when KubeStellar supports the use of the full Kubernetes API for management of containerized workloads (today only a small subset of those, all namespaced, is supported: Deployment, Pod, Service, Ingress).

Dependency 2 will soon merge into Dependency 1, because KubeStellar will adopt those resources.

The work on an abstraction layer for the logical cluster functionality should remove Dependency 3.

Following are some approaches for dealing with Dependencies 1 and 4.

Approach A (for Dependency 1 only)

Change the KubeStellar API resources to be namespaced.  Define one particular namespace that they belong in, or allow users to organize those objects into namespaces in whatever way the users want.

Approach B (for Dependency 4 only)

Let the logical clusters have the full Kubernetes API built-in.  Not the corresponding controllers, just the resource definitions.

Approach C (for dependency 1 and/or 4)

(Definition: the kube-bind syncer propagates desired state from _consumer-side objects_ to their _provider-side copies_ and reported state from those provider-side copies to the corresponding consumer-side objects).

Extend the kube-bind syncer so that it can propagate cluster-scoped objects.  Because the syncer is not sensitive to the semantics of the objects, it can have generic code that operates on any configured resources.

The biggest question I see here is what to do about name collisions.  The provider-side copies can not simply have the same name as the consumer-side objects.  Making a provider-side copy's name contain the original name plus the consumer name can run into problems with both length and charset/syntax checking on the names.  The provider-side copy's name could be arbitrary generated anything, but where to stash the original name?  The content and length of a label value is too constrained to be a viable solution.  That leaves annotations.  The down-side of that is that they can not be tested in the filter of LIST and WATCH requests.  I do not yet see how that would be a practical problem.

Approach W (for Dependency 1 and/or 4)

Define a special resource that appears on the provider side and wraps the provider-side copies.  The wrapper objects are namespaced, each appearing in the provider-side namespace corresponding to the consumer cluster.

**Comment by pdettori** [comment]
From an API/UX (and not implementation) perspective I think approach B or C would be preferable. For B, KubeFlex "logical clusters" already satisfy that property (full Kubernetes API built-in. Not the corresponding controllers)

**Reply by clubanderson** [reply]
I like B. It seems the cleanest and leanest approach. Consistent as well

**Reply by MikeSpreitzer** [reply]
Approach B addresses Dependency 4 but not Dependency 1.  So we would need one of A, C, or W for Dependency 1.

**Comment by waltforme** [comment]
I was making some efforts in the past few days, which fits into Approach C If my understanding is correct.

I was digging into kube-bind and found a four-tier structure of the consumer side controllers, towards the 'syncer':
1. The konnector, which starts a cluster controller;
2. The cluster controller, which starts four controllers including the serviceexport controller;
3. The serviceexport controller, which dynamically starts a pair of controllers - spec and status controllers - for one specific GVR, or in other words one specific apiserviceexport, or in other words one specific CRD;
4. A pair of controllers to sync spec and status between a consumer and a provider.

Going forward with Approach C, we need two more steps to export EdgePlacement.
1. We need somehow 'enable' the GVR for EdgePlacement;
2. We need to change current syncer (the pair of spec and status controllers) because the current syncer focuses on namespaced APIs.

One concern (in addition to the name collision issue) of Approach C I have is that, we may have to maintain our own fork of kube-bind, but not merging our work into upstream (kube-bind/kube-bind). Because it seems that kube-bind tends to use kcp to handle cluster scoped APIs which is conflicting with Approach C.

BTW, when I was digging, I found some mistakes and typos in the code base. I submitted a PR to kube-bind: https://github.com/kube-bind/kube-bind/pull/165

**Comment by MikeSpreitzer** [comment]
@waltforme: thank you.  I do not understand your statement that "it seems that kube-bind tends to use kcp to handle cluster scoped APIs".  I thought kube-bind works between ordinary Kubernetes clusters and does not rely on the kcp server.

I was thinking that if we take approach C then this is something that we could contribute to kube-bind itself rather than maintain a fork.

Also, I noticed that kube-bind's `go.sum` imports `pkg/apis` from v0.11.0 of kcp.  What is kube-bind using from the v0.11.0 kcp apis, and where?

**Reply by waltforme** [reply]
> Also, I noticed that kube-bind's `go.sum` imports `pkg/apis` from v0.11.0 of kcp. What is kube-bind using from the v0.11.0 kcp apis, and where?

I checked this weeks ago. It turns out that kcp and logicalcluster/v3 are used merely in the e2e test.

**Comment by waltforme** [comment]
> @waltforme: thank you. I do not understand your statement that "it seems that kube-bind tends to use kcp to handle cluster scoped APIs". I thought kube-bind works between ordinary Kubernetes clusters and does not rely on the kcp server.
> 

My understanding is that
- kube-bind could use either vanilla k8s or kcp ([not implemented yet](https://kubernetes.slack.com/archives/C046PRXNJ4W/p1686066422321759?thread_ts=1686065450.317109&cid=C046PRXNJ4W)) as the provider, shown as the headline of the 'provider' box in the diagram;
- if k8s is used as the provider, cluster-scoped APIs are not supported, so in the diagram, 'ClusterScoped Resource' is marked 'only kcp';
- in other words, to support cluster-scoped APIs, it is planned to use kcp as the provider.

Surely we want to contribute Approach C to upstream, but since kcp and APIExport+APIBinding is planned as the approach to handle cluster-scoped APIs, I'm afraid that there will be the aforementioned conflict, and we will end up with a fork.

![image](https://github.com/kcp-dev/edge-mc/assets/8633434/9911f141-e5ab-40d4-843a-dd6a925786b1)

**Reply by MikeSpreitzer** [reply]
I think kube-bind should aim to be usable with plain Kubernetes providers, so contributing approaches C and/or W upstream should be acceptable.
